### ISSUE TITLE
iOS + iPadOS Support

### ISSUE THREAD

**Issue by Kartik-715** [quoid/userscripts#135]
With the release of iPadOS and iOS 15, do you have plans to migrate this extension to a safari web extension? Could be pretty useful!

**Comment by quoid** [comment]
@Kartik-715 

I have to read more about the support. It *seems* like Apple is supporting Safari WebExtensions in iOS, not Safari *App* Extensions. This extension is an App Extension not a Web Extension.

If that is the case the extension would need to be re-written and current functionality could be lost. At the current time I do not know.

**Comment by quoid** [comment]
@Kartik-715 It is possible it's just a matter of time and rewriting certain functionality like interacting directly with the file system. I don't have any doubts it is possible.

I am waiting until more information is out about the new iOS 15 features before I make any plans. However, once I hear more I will likely be re-writing, so I would suggest not to waste your time attempting to migrate anything. If you want to help when I begin the rewrite that would be more beneficial to the project.

**Comment by Reinachan** [comment]
> You implement Safari web extensions as macOS or iOS app extensions to provide a safe and secure distribution and usage model. You can distribute a Safari web extension with a Mac app, an iOS app, or a Mac app created using Mac Catalyst. You must use Xcode to package your extensions for testing in Safari, and you must be a member of the Apple Developer Program to distribute them through the App Store. For more information, see Apple Developer Program.
https://developer.apple.com/documentation/safariservices/safari_web_extensions

Looks like Safari App Extensions and Safari Web Extensions will work on iOS?

**Comment by quoid** [comment]
> Looks like Safari App Extensions and Safari Web Extensions will work on iOS?

I read the quote, but can you elaborate on why you think Safari *App* extensions will work? Specifically app extensions without WebExtension components.

**Comment by Reinachan** [comment]
Ah, without web extension components. Well, then I wouldn't know. I added a questionmark to it as the wording by Apple was a bit vague.

Looking deeper into the various documentation pages, it looks like it's Safari Web Extensions but with the ability to have a native app to control it from.

It also looks to me like they're moving away from Safari App Extensions in general and toward Web Extensions instead. Might be worth considering refactoring to Web Extension either way, as they might just end up cutting support for Safari App Extensions down the line. Who knows.

Moving to Web Extensions could also have the advantage of making it much easier to enable using the extension in other browsers too from what I can gather?

**Comment by quoid** [comment]
@Reinachan 

Thanks for the clarification. I agree that migrating to Web Extension will prove valuable and that is my goal, however I can not give a firm estimate on when that will be complete as it is will take some time.

That being said, simply converting to being *predominately* WebExtension API won't instantly enable usage in iOS as there are other aspects of the extension to consider, such as selecting the "scripts folder" and mobile considerations for the extension page (ie. editor which will likely barely function on any mobile device currently).

> It also looks to me like they're moving away from Safari App Extensions in general and toward Web Extensions instead. Might be worth considering refactoring to Web Extension either way, as they might just end up cutting support for Safari App Extensions down the line. Who knows.

Once they announced WebExtensions, I had the same thought. It's already sort of happening since the WebExtension API is much more robust compared to the App Extension API (which gets seldom updates). It does slightly irk me that Apple decided to drop Safari JS Extension API and force users into App Extension API only to (a few years later) seemingly begin abandoning App Extension API in favor of WebExtension API. It's better for the browser but reveals their finicky and haphazard forward thinking (in terms of Safari).

> Moving to Web Extensions could also have the advantage of making it much easier to enable using the extension in other browsers too from what I can gather?

Very true, however at the current moment I don't have much interest in adding additional browser support since there are really good options for those browsers. But that opinion could change in time.

Anyway, WebExtension refactor is planned, stay tuned!

**Comment by Reinachan** [comment]
You've got some good points in regards to an iOS port. I think that for an initial release, simply giving the option to add/remove userscripts/styles you find on the web could be enough, and potentially expand upon it later.

Personally, I just want to use the userstyles I write on my PC on iOS for instance, and I will assume that a majority of people who use plugins like this don't write them themselves.

I don't know what APIs they give you access to, but considering they let you add a native app to control certain stuff in the extension, perhaps you also get access to a folder in the filesystem which you can store things in? In that case, you might not even need some way to write them, as users could just write them in their preferred program and add it to that folder?

I'd contribute to this but my Macbook decided that it no longer wants to boot regularly (only if in safemode), so I had to install Linux on it to even be able to continue using it. Means I won't really be able to test anything or deploy it to my iPad :/

**Comment by quoid** [comment]
> I don't know what APIs they give you access to, but considering they let you add a native app to control certain stuff in the extension, perhaps you also get access to a folder in the filesystem which you can store things in? In that case, you might not even need some way to write them, as users could just write them in their preferred program and add it to that folder?

I think that would exactly what I'd want to see in a v1 for iOS - simply the ability to read userscripts from a directory of the user's choosing. For example, I store all my scripts in an iCloud folder, if I am able to read from that directory and utilize the userscripts located there, on my phone, I would be very pleased.

Long term possibilities are endless, but I am fairly certain that creating an editor for iOS or iPadOS is more than likely **not** a goal.

> I'd contribute to this but my Macbook decided that it no longer wants to boot r

You already are contributing by sparking up conversation and it's valuable!

**Comment by quoid** [comment]
track progress here:

#145 

https://github.com/quoid/userscripts/milestone/6

https://github.com/quoid/userscripts/tree/v4.0.0

**Comment by julianpomper** [comment]
@quoid I think I found your issue. The problem is that the `--rebuild-project` option is just available with the Xcode 13 developer tools. So if you run `xcrun` but did not change the active development directory to the Xcode 13 tools it does not find this option.

You can try entering `sudo xcode-select path/to/Xcode.app` (path to Xcode 13, you can change it back later). Now the command `xcrun safari-web-extension-converter --rebuild-project /path/to/project.xcodeproj` should work 👍

**Comment by quoid** [comment]
@julianpomper I had tried that in the past and just tried it again this morning and the results are the same. The converter does not do anything but print out help text. Interestingly the help text does not state that `--rebuild-project` is a valid option.

Have you successfully rebuilt  an existing **macIS** webextension?  If so, what OS are you on and Safari version? I am using Safari `14.1.1` and macOS `11.4` and just figured the tool wasn't available fully until the next macOS.



<img width="825" alt="Screen Shot 2021-09-17 at 8 57 48 AM" src="https://user-images.githubusercontent.com/7660254/133786389-32efb5ac-6638-4851-a6a5-fb0ad94aca04.png">

**Comment by quoid** [comment]
@julianpomper 

> option is just available with the Xcode 13 

I re-read your post and realized I am on Xcode 12... Sorry it is early for me 🙂 

I reckon I won't be able to covert and build the iOS side until I am on the new macOS. When that gets released I should be able to update and run the convertor and start that process.

**Comment by julianpomper** [comment]
Exactly you should already be able to do that now with macOS 11.4. Download the Xcode 13 RC [here](https://developer.apple.com/download/) in the `Beta` tab. After that open it once and install the additional tools. If that is done you should be able to do the steps I described above:

> You can try entering sudo xcode-select path/to/Xcode.app (path to Xcode 13, you can change it back later). Now the command xcrun safari-web-extension-converter --rebuild-project /path/to/project.xcodeproj should work

**Comment by quoid** [comment]
@julianpomper I just wanted to give you an update. The good news is that I was able to convert the extension to iOS 🎉 

The bad news is that a **good deal** of the [v4.0.0](https://github.com/quoid/userscripts/tree/v4.0.0) code needs to be re-written `#if os(iOS)`. This will likely take a long time unless I can find an experienced iOS developer to help with this.

**Comment by EthanG45** [comment]
What do you mean rewritten? Coming in here with no context just interested in seeing any user script app reach iOS. Do you mean the scripting code no longer works or just UI is broken?

**Comment by quoid** [comment]
> What do you mean rewritten?

@EthanG45 

Here's some examples:

- Currently the app uses security scoped bookmarks, which are set by the user through the containing app, to set the read/write location. The way this is written in macOS is not comptatible with iOS. 
- By default the app reads from the users `.documentsDirectory`, but for the iOS version there'd have to be a specific workflow which compels the user to set a shared folder as a read location. Creating userscripts is not in the scope for the iOS version.
- When converting the extension is does not keep any of the swift functionality that is included in the current and beta versions. That needs to be integrated back in after the conversion, or better yet, the conversion needs to happen, then that converted version needs to be integrated into the current project (or some similar approach).

Anyway, my point is there's a lot of stuff that needs to get to done to even figure out if it is viable to run on iOS and iPadOS. Some of that includes iOS development which I am inexperienced with. If all of this is left up to me, it will *probably* get done, but just take *quite* a while.

**Comment by EthanG45** [comment]
Ah I see that is a lot of non-easily converted code. I've got some naive iOS experience mostly on the UI side, so I can't say I could contribute much, at least right now I've got a full plate.  Hopefully, one day userscripts can come to iOS still.

**Comment by quoid** [comment]
> Hopefully, one day userscripts can come to iOS still.

It might take a little bit but I think it will get there @EthanG45

**Comment by quoid** [comment]
**Update**

I've been running some tests. Outside of adapting current code to work on iOS, I am still in the process of figuring out if choosing a read location can persist between Safari launches. Choosing a location for reading the userscripts will have to be done as it is currently done in the current macOS app. That means the user selects this location in the containing app, in this case the iOS app and then a sharedbookmark is created and stored in a shared userdefaults file. 

That shared userdefaults value is then read by the extension. In the simulator, it seems like this persists, but I am unsure how this will function on an actual phone/ipad. I've read some random forum posts suggesting that it is not possible to have sharedbookmarks persist as they can in macOS. **If that is true, I am not sure how to approach all of this.** However, it seems to persist currently, so I am hoping it is *not* true.

Once that is figured out, I think the path is clear. I'm successfully running this in the simulator currently.

<img width="993" alt="Screen Shot 2021-10-02 at 11 01 48 AM" src="https://user-images.githubusercontent.com/7660254/135722320-c39d1d36-72ad-4388-9c11-10a44283c9e1.png">
<img width="993" alt="Screen Shot 2021-10-02 at 11 02 05 AM" src="https://user-images.githubusercontent.com/7660254/135722321-fea0e2be-64a5-4566-9d93-f01666ecb24c.png">

**Comment by EthanG45** [comment]
@quoid That's fantastic to hear! Looking forward to future updates on development progress.

**Comment by quoid** [comment]
**Update**

I updated my iPhone 8 to ios 15 and loaded the extension. Injection seems to occur, will update everyone on if the bookmark persists. However, the popup does not load properly. This is not something I experienced in the simulator.

I made a thread in the Apple forums in hopes of deciphering the error code: https://developer.apple.com/forums/thread/691389

If anyone has any insight please post here

**Comment by quoid** [comment]
**Update**

The iOS app has been on my phone for a bit over a day and is still working. The popup still randomly fails, but subsequent reloads after a failure seemingly resolves it. I can't figure out what's causing it to fail so I opened a dev support ticket with Apple.

I will be looking into setting up Test Flight this weekend so I can get more hands on this to help out with testing.

**Comment by RedGuy12** [comment]
Thank you for your work in maintaining this extention (and especially making it free :P)


________________________________

**Comment by EthanG45** [comment]
Awesome! Definitely share the TestFlight as I can put it through its paces with some of my scripts.

**Comment by quoid** [comment]
> Just throwing it out there, but have you tried the iOS 15.1 beta? Might be an already fixed OS/Safari issue causing it.

Thanks for the suggestion @NewsGuyTor, I haven't installed it, but that would be great if it did. Hopefully I can get this Testflight stuff figured out soon so I see if other people experience it as well.

**Comment by a9eelsh** [comment]
Hello can use https://apps.apple.com/sa/app/hyperweb/id1581824571 
support userscript

**Comment by quoid** [comment]
I just deleted a comment that was promoting a paid app that someone was promoting. 

I don't think promoting paid apps is relevant to this issue. If it were discussing another open source project that could help with the development of this feature the comment would remained.

**Comment by madman404** [comment]
I feel at the mention of "that app", I should further emphasize how _much_ that **isn't** helpful by emphasizing the reason open-source projects like this are good for the community to begin with. I'd like to use userscripts without as much device information as can be gathered being collected and **tied to my unique IP address** as possible.

**Comment by willl** [comment]
> I'd like to use userscripts without as much device information as can be gathered being collected and **tied to my unique IP address** as possible.

It also seemed to do a bit of "everything" rather than an extension that allowed the user to make use of userscripts. 

Anyway... Back on topic, I look forward to testing this extension whenever it is made available.

**Comment by a9eelsh** [comment]
Hey bro, any news?

**Comment by quoid** [comment]
@a9eelsh 

> Can you give existing here IPA for quick experience?

I am afraid I don't know what that is. I'm not an iOS developer so this is all new to me.

**Comment by a9eelsh** [comment]
https://wiki.genexus.com/commwiki/servlet/wiki?34616,HowTo%3A+Create+an+.ipa+file+from+XCode Example

**Comment by a9eelsh** [comment]
https://wiki.genexus.com/commwiki/servlet/wiki?34616,HowTo%3A+Create+an+.ipa+file+from+XCode Example

**Comment by quoid** [comment]
@Kartik-715  @willl @madman404 @Reinachan @julianpomper @EthanG45 @RedGuy12 @a9eelsh and anyone else viewing this thread. Version 0.1.0 beta for iOS got approved for testflight.

If you want to test, please sign up here: https://forms.gle/QB46uYQHVyCxULue9 - don't forget to use the email associated with your Testflight account. **Even if you signed up in the past to beta test, you need to re-sign up because I just purge the old tester pool.**

I reckon this first version will be rough, so please adjust expectations. At the current time I am mainly looking for **iPhone testers** running iOS 15+. The iPad version doesn't really work at the moment (visual bugs with the popover).

**Comment by bradenglund** [comment]
I was able to install and enable as expected. Unfortunately I can’t get past the “Something went wrong - click to retry” message despite force closing safari, app, and disabling/re-enabling extension.

I have the userscripts folder saved on iPhone, not iCloud if that matters.

**Comment by quoid** [comment]
@bradenglund 

That is the issue I commented on above. I was hoping that others would not experience it as much as I did. I know other people are also experiencing it, so I reckon everyone who currently is testing is having the same problem.

The core of the problem is that during the init phase of the popup, an error occurs that I can find the origin of. I added that "Something went wrong" message to try and deal with it. I used one of my Apple Code-Level support requests but the engineer thinks it's a bug and told me to file a bug report, which doesn't really help.

In my situation, I chose a different directory than my main userscripts directory on my icloud drive and once that one worked, I subsequently selected my main userscripts directory again, and the error did *always* occur. It still sometimes happens, but I am able to refresh the popup (by clicking retry) or by clicking the "refresh" button in the top of popup and it usually goes away.

I am going to try and figure out a way to get around this and push a new build asap.


<img width="848" alt="Screen Shot 2021-10-02 at 10 08 07 PM" src="https://user-images.githubusercontent.com/7660254/137591661-78e716ed-36e9-45d9-aa80-63263f1b6d1c.png">

**Comment by quoid** [comment]
I just pushed a new build, I expired the previous. This fixes the ui on my iPad and hopefully others, so that should be able to be tested now.

I also include a numeric "error code" that prefixes the `something went wrong` message, in an attempt to track down what is going on.

Here are some things to try if you encounter the `something went wrong` message:

- refresh the popup by clicking the “click to retry” link
- refresh the popup by clicking the “refresh icon” (icon that looks like a circular arrow)
- closing the popup and reopening it
- closing safari and reopening
- trying to open the popup on a different page
- selecting a folder with fewer userscripts in it, then switching back to the main userscripts folder


If you are able to successfully run the extension, **please let me know**. I am not seeing the issue persistently on my iPhone, however my iPhone is "used for development" so I am unsure if that makes the issue less prevalent.

**Comment by bradenglund** [comment]
> I chose a different directory than my main userscripts directory on my icloud drive and once that one worked, I subsequently selected my main userscripts directory again, and the error did not always occur. It still sometimes happens, but I am able to refresh the popup (by clicking retry) or by clicking the "refresh" button in the top of popup and it usually goes away.


I tried this, but am still experiencing the error every time. It only works in a blank directory and then says none found. Otherwise always error.

**Comment by bradenglund** [comment]
On new build I a getting error “3 Something went wrong”

On the bright side it appear to function on refresh, on/off and update for a blank directory now.

**Comment by v7rzlhb3xg** [comment]
Same error, "3 Something went wrong." 

Interestingly, after installing it on my iPad, I wasn't getting that error for the first couple minutes. However, it either wasn't able to find any of the scripts in the folder I pointed it to, or it wasn't matching it to the website. Simply put, no scripts were showing up. It wasn't until several minutes later that I got the previous error, though. That being said, the iPad UI works great!

**Comment by v7rzlhb3xg** [comment]
Yes, the same error persists even after refreshing, changing the folder, restarting safari.

**Comment by quoid** [comment]
> Yes, the same error persists even after refreshing, changing the folder, restarting safari.

:(

Ok, I am going to try some more things. Possibly a new build today. There's something that's screwing up the native messaging. I also noticed a random `Couldn't Communicate with a helper application` error when testing, which I think is related to the Files.app.

**Comment by bradenglund** [comment]
> Yes, the same error persists even after refreshing, changing the folder, restarting safari.

Same true for me.

**Comment by EthanG45** [comment]
> @Kartik-715  @willl @madman404 @Reinachan @julianpomper @EthanG45 @RedGuy12 @a9eelsh and anyone else viewing this thread. Version 0.1.0 beta for iOS got approved for testflight.
> 
> 
> 
> If you want to test, please sign up here: https://forms.gle/QB46uYQHVyCxULue9 - don't forget to use the email associated with your Testflight account. **Even if you signed up in the past to beta test, you need to re-sign up because I just purged the old tester pool**, to prevent sending unwanted Testflight invites.
> 
> 
> 
> I reckon this first version will be rough, so please adjust expectations. At the current time I am mainly looking for **iPhone testers** running iOS 15+. The iPad version doesn't really work at the moment (visual bugs with the popover).

Already filled all slots?

**Comment by a9eelsh** [comment]
Can I try the app?

**Comment by quoid** [comment]
@EthanG45 

There's a **major** issue that is preventing, I assume, all users from being able to interact with the popup. Because of this I've closed sign ups until it becomes stable.

I've been digging into this all morning trying numerous work arounds and nothing seems to be working. I am guessing here, but I think the issue could be:

- the extension is hitting the very small memory limit allows in the current version of iOS (currently 6mb, but in 15.1 it will be 80mb) - unfortunately I do not know how to even test how much memory is being used by the extension so I can't verify if we are at the limit
- perhaps some issue with the front-end frame work - I am considering rewriting the popup in vanilla js/html/css to test

**Comment by quoid** [comment]
I just pushed build 4, in a possibly futile attempt to pin down this issue. I disabled remote fetching as I think I was noticing a pattern with those functions and the issue ¯\_(ツ)_/¯

@EthanG45 shoot me an email if you want to give it a shot (its in my profile)

**Comment by v7rzlhb3xg** [comment]
Still getting “Something went wrong” when pointing to a directory with .js files but no number this time. Setting it to any other directory lets me get farther into the update UI than I was able to before though! 
![35DBFC3B-9C74-4F33-AA2F-B312C498D2B7](https://user-images.githubusercontent.com/18043024/137603999-6b5db3a9-e5d2-466c-9db7-a451968ae879.png)

**Comment by quoid** [comment]
@a9eelsh if you want, but I am not sure it will work ^^

@v7rzlhb3xg Damn. I installed it through Testflight on a friends iPad 6th gen that has no association with my dev account and was having the same issues as everyone else. Once I disabled remote fetching, I didn't experience the issue any longer. I did however plug the ipad into a macbook (again no association with my dev account) and inspected the popup through Safari "develop" menu to check the console.

Do you have any "undownloaded" files in your Files.app folder?

**Comment by a9eelsh** [comment]
> 
> 
> @a9eelsh if you want, but I am not sure it will work ^^
> 
> @v7rzlhb3xg Damn. I installed it through Testflight on a friends iPad 6th gen that has no association with my dev account and was having the same issues as everyone else. Once I disabled remote fetching, I didn't experience the issue any longer. I did however plug the ipad into a macbook (again no association with my dev account) and inspected the popup through Safari "develop" menu to check the console.
> 
> Do you have any "undownloaded" files in your Files.app folder?

yse i need try this

**Comment by v7rzlhb3xg** [comment]
@quoid Great news! No more error! I believe you were right the first time -- it's absolutely a memory issue. There was a rather large userscript (1.1MB) in the original folder I've been trying to use. After trying with a new folder and a smaller test user script (9KB) the error is gone. I don't believe whether I had undownloaded files in my iCloud downloads folder doesn't seem to make a difference.

**Comment by quoid** [comment]
@v7rzlhb3xg 

Interesting! We are making progress!

I am still unsure if it is a memory issue or something to do with remotely fetching code/data, or perhaps both are related. 

As mentioned in previous comments, I almost completely stopped experiencing the issue on my personal iPhone. I thought that perhaps because that was associated and used with my developer account something made it less prone to the issue. However, this afternoon I was able to reproduce the issue again on my phone.

I noticed that during the popup init phase, when I was checking for required resources (`@require` key in userscripts), the error would appear. It seemed related to the remote data fetching I was doing. This most recent update disabled all of that, which is not optimal for version 1 of the iOS app, but gives me some insight into what might be causing the issue.

I am curious if the userscript you removed relied on the `@require` key or `@updateURL` key.

Version 3 is still available, so if you want to try re-installing that version and seeing if the error reappears, even though you removed that file, that would be super helpful 😄

**Comment by v7rzlhb3xg** [comment]
@quoid I can't seem to find Version 3 under previous builds in TestFlight, only the latest Version 4 is showing up for me. 

In the meantime, I did some more testing. The [large script](https://www.4chan-x.net/builds/4chan-X.user.js) that was causing problems did in fact have `@updateURL` but still throws the error even with that removed. Deleting that script from my userscript folder resolves the issue again. Maybe there's something else in there that's problematic?

**Comment by EthanG45** [comment]
@quoid I just installed the beta on my iPhone 13 Pro Max and iPad Pro 2018 12.9" both running iOS/iPadOS 15.0.2. There were some UI bugs, but it loaded up my scripts folder from my macOS version of user scripts and correctly loaded in my FrankerFaceZ script for Twitch. Will continue testing but this seems to be working so far!

**Comment by bradenglund** [comment]
> @quoid Great news! No more error! I believe you were right the first time -- it's most likely a memory issue. There was a rather large userscript (1.1MB) in the original folder I've been trying to use. After trying with a new folder and a smaller test user script (9KB) the error is gone. I don't believe whether I had undownloaded files in my iCloud downloads folder seemed to make a difference, but I emptied it just in case.

I also was able to get it to load by using a very small single file in a directory by itself.
https://openuserjs.org/scripts/mscarchilli/Amazon_Smile_Redirect
![E42EA60D-0A3A-4E54-B95E-4E36133C2215](https://user-images.githubusercontent.com/28195153/137604794-54635681-83a0-4795-b296-6fe30d0ae565.png)

It is not actually executing the script as I am still on www.amazon.com, but it's a start.

Unfortunately, if I switch directories back to the one with several scripts, "Something went wrong" with no numerical code. I also tried another single js file, it fails for that too. For reference, this is another I am testing with.
https://github.com/SteppoFF/ita-matrix-powertools/

**Comment by bradenglund** [comment]
I was also able to reproduce the results of the amazon smile script with this script. https://github.com/hjk789/NXEnhanced/tree/master/Userscript

**Comment by quoid** [comment]
@v7rzlhb3xg 

>  Deleting that script from my userscript folder resolves the issue again. Maybe there's something else in there that's problematic?

I'm still unsure what the actual cause is. I was thinking that maybe it was remote file content fetching, but maybe that isn't the actual cause. Maybe you (and me) were getting that issue because when requesting remote file contents, the pay load made the extension exceed the memory limit (this is just a guess).

For my userscripts, I have 2 in particular that request large-ish libraries (~100kb), so I am wondering if that could be it.

Either way, it's clear more testing is needed all around.

**Comment by quoid** [comment]
After many hours spent today I still have no clue what is causing these errors. I just did a video call with a friend and we went from fresh install to usage and while we encountered the issue, overall the extension seemingly worked as intended (and this was with a build that re-enabling updated and remote fetches).

We did have to jump between selected folders, one with many scripts another with only, but not perceivable failure pattern could be observed. When we ended the call, the folder that was originally giving us `something went wrong` was working.

I'm at a real loss as to what is going on here, but will continue to investigate. I truly appreciate everyone who is helping out with the testing.

**Comment by willl** [comment]
Can confirm similar issues. When I cleared out some larger scripts, it seemed to work fine.

The simple userscript you posted worked fine on both iPhone and iPad. 

Haven't had time to try other scripts yet, but will do when I have time.

**Comment by a9eelsh** [comment]
> > @quoid Great news! No more error! I believe you were right the first time -- it's most likely a memory issue. There was a rather large userscript (1.1MB) in the original folder I've been trying to use. After trying with a new folder and a smaller test user script (9KB) the error is gone. I don't believe whether I had undownloaded files in my iCloud downloads folder seemed to make a difference, but I emptied it just in case.
> 
> I also was able to get it to load by using a very small single file in a directory by itself. https://openuserjs.org/scripts/mscarchilli/Amazon_Smile_Redirect ![E42EA60D-0A3A-4E54-B95E-4E36133C2215](https://user-images.githubusercontent.com/28195153/137604794-54635681-83a0-4795-b296-6fe30d0ae565.png)
> 
> It is not actually executing the script as I am still on [www.amazon.com](http://www.amazon.com), but it's a start.
> 
> Unfortunately, if I switch directories back to the one with several scripts, "Something went wrong" with no numerical code. I also tried another single js file, it fails for that too. For reference, this is another I am testing with. https://github.com/SteppoFF/ita-matrix-powertools/

I tested it and it works fine for me

**Comment by a9eelsh** [comment]
Update 1 

i’m try two scripts 

Script 1
`function betterttv() {
    const script = document.createElement('script');
    script.type = 'text/javascript';
    script.src = 'https://cdn.betterttv.net/betterttv.js';
    const head = document.getElementsByTagName('head')[0];
    head.appendChild(script);
}

betterttv(); `

Script 2
`(function betterttv() {
    var script = document.createElement('script');
    script.type = 'text/javascript';
    script.src = 'https://cdn.betterttv.net/betterttv.js';
    var head = document.getElementsByTagName('head')[0];
    if (!head) return;
    head.appendChild(script);
})() `


The first works fine, but second doesn't

**Comment by quoid** [comment]
@a9eelsh 

> The first works fine, but second doesn't

Change second code to:

```js
;(function betterttv() {
    var script = document.createElement('script');
    script.type = 'text/javascript';
    script.src = 'https://cdn.betterttv.net/betterttv.js';
    var head = document.getElementsByTagName('head')[0];
    if (!head) return;
    head.appendChild(script);
})();
```

**Comment by bradenglund** [comment]
> 
> 
> @bradenglund
> 
> > It is not actually executing the script as I am still on [www.amazon.com](http://www.amazon.com), but it's a start.
> 
> I downloaded the first Amazon Smile script and it was executing, but the code was made for the desktop site. This element: `document.getElementById('nav-tools');` is required for the script to redirect and it doesn't exist on the mobile site.
> 
> > with no numerical code
> 
> I removed the numerical code in the latest build as it wasn't as helpful as I hoped. The fact that you still see the error makes me think that it is not remote fetching but more so a memory issue, but more testing needed to verify.
> 
> Here's a simple userscript for testing :
> 
> ```js
> // ==UserScript==
> // @name        iosTest
> // @description only for testing on ios
> // @include     https://www.k21p.com/
> // @run-at document-start
> // @noframes
> // ==/UserScript==
> 
> alert("iosTest");
> ```

So I was able to test with this and it works fine in a directory by itself. It also works if I have _either_ one of these two scripts in the same directory, but **not both**.
https://github.com/AdguardTeam/PopupBlocker
https://openuserjs.org/scripts/mscarchilli/Amazon_Smile_Redirect

I played around with a couple of the other scripts, and those all caused `Something went wrong` as well.

**Comment by quoid** [comment]
@bradenglund 

Thanks for letting me know. I am still lost. Perhaps it has to do with file size and a memory limit?

Yesterday @kfatehi and I were able to do a fresh install on his iphone with all features enabled (`@require`, `@updateURL`, etc...) and we did encounter the `something went wrong` error, but were able to circumvent it by juggling the read directory. At one point he had 22 userscripts in the directory with no errors.

The errors just seemed random. For example, at one point, anytime we put a specific userscript in the folder, the error would appear - so we removed it. Then when we put that userscript back in the folder, the error did *not* appear. We gained no valuable clues as to what was going on.

For me, most of my userscripts are self-made and quite succinct, which I am wondering why I have not experienced many issues. Whereas userscripts from outside authors tend to be larger to accommodate many users.

So right now I feel no closer to figuring this out than I did a week ago, but I think there are ways to get closer.


### If anyone can help figure out the source of this error: https://developer.apple.com/forums/thread/691389

### If anyone knows how to monitor memory usage in Safari Web Extensions, please let me know

### If anyone has experience with iOS development and is willing to do a code review, email me

### If you are encountering the error and want to screen share to look patterns, email me


Outside of that I will keep throwing darts

**Comment by quoid** [comment]
We have good and bad news.

The good news is that **I found the problem!!!** I am mad I missed this earlier. It took hours of looking through the logs and it was right under my nose.

The bad news is that until iOS 15.1 is released there will be issues.

It's a memory limit issue:
<img width="933" alt="Screen Shot 2021-10-17 at 2 09 12 PM" src="https://user-images.githubusercontent.com/7660254/137639593-42ff11aa-1807-4086-9b49-324f68eeb227.png">

This why is works sometimes and does not for others times.

I guess we have a few weeks to wait.

¯\_(ツ)_/¯

**Comment by v7rzlhb3xg** [comment]
Amazing news! Userscripts has been working flawlessly for me with the couple small scripts I’ve been using. Really impressed, this is an absolute game changer for my iPad. Thank you so much for all of your hard work!

**Comment by Kartik-715** [comment]
Hi @quoid,
Have we run out of the slots for the number of people who can test? I filled the form yesterday and still haven't got the email to sign up.

**Comment by quoid** [comment]
> Hi @quoid,
> Have we run out of the slots for the number of people who can test? I filled the form yesterday and still haven't got the email to sign up.

@Kartik-715 

I stopped sending out the invites because of the memory issues. The extension kind of works depending on how many scripts there are but it's not very reliable at the moment. 

If you still want me to add you I will but keep that in mind if it doesn't function well. Just reply here if you want me to send it.

**Comment by quoid** [comment]
@cynix I re-opened the form, although unless I am specifically asked (like you did) I won't be sending out invites to people until 15.1 is released.

If you signed up and email me (email in profile), I will add you.

**Comment by elh-rotem** [comment]
Just tried to do a simple alert on YouTube, it does work, but not if YouTube is opened as a Home Screen shortcut. Seems like a bug to me.

**Comment by quoid** [comment]
@elh-rotem please post the userscript (even if simple) and steps to reproduce, including device, and os version. I'm not sure what a Home Screen short cut is.

**Comment by elh-rotem** [comment]
`// ==UserScript==
// @name        iosTest
// @description only for testing on ios
// @include     https://www.youtube.com/
// ==/UserScript==

alert("iosTest"); `

Using iOS 15.1 on iPad Pro 11” 2021. There’s an option on Safari to add to the Home Screen a shortcut to a specific site.

**Comment by quoid** [comment]
@elh-rotem 

> `// @include https://www.youtube.com/`

If you're visiting youtube on an iOS device via Safari, iPad or iPhone, you are redirected to `https://m.youtube.com`. You'd have to update that `@include` value.

**Comment by elh-rotem** [comment]
@quoid Tried what you’re suggesting, Issue still persist.

**Comment by quoid** [comment]
@elh-rotem 

I can not replicate the issue you describe when the error that was in your original userscript is corrected. I suspect there's an issue with how the userscript was coded.

https://user-images.githubusercontent.com/7660254/140205098-eaef6a03-3a3f-42c0-a76a-5e91801d0e40.MP4

**Comment by EthanG45** [comment]
@quoid TestFlight for the Mac just released and looks like Userscripts iOS is accidentally(?) marked as available for Apple Silicon Macs. Just an FYI.

<img width="822" alt="image" src="https://user-images.githubusercontent.com/55003415/140273404-5b9c7010-062b-41db-a78e-ea3dcbddb16a.png">

**Comment by quoid** [comment]
Thanks for letting me know @EthanG45, that's certainly not purposeful. I will see what I can do later today to correct that.

**Comment by quoid** [comment]
I will be dispatching a new version, which resets the build numbers soon. The reason for this change is that the first app distributed through Testflight was called "Userscripts iOS" because I had not "associated" with the macOS version. However, I corrected that by creating an iOS version of the macOS and both are using the same bundle id now. Everyone testing will soon get a possibly confusing invite to test "Userscripts" for iOS which should be accepted if you want to continue testing.

I will be expiring the first version(s) that was sent out shortly after sending the invites for the new version.

I apologize for the confusion.

PS - I am working on getting v4.x of the macOS app on Testing as well. Hopefully that beta will be set up through Testflight soon enough.

**Comment by quoid** [comment]
Also, @cynix - the semicolon issue is fixed in the latest build

**Comment by vjjj4688** [comment]
I tried to download several scripts in the happy fork, but these scripts can't be used. I don't know why, the scripts can only be seen in the top extension, but they can't take effect

**Comment by quoid** [comment]
@vjjj4688 

> I tried to download several scripts in the happy fork

What is happy fork?

Is you having an issue with the new version of the app or the old? This is easy to tell by the current build number, which is shown in the iOS app, example -> `v0.1 (x)`

**Comment by vjjj4688** [comment]
sorry，https://greasyfork.org/，I mean this script download website，I am using the latest version of 0.1 (2)，It doesn't work like previous versions

**Comment by quoid** [comment]
@vjjj4688 

Thanks for the clarification.

> It doesn't work like previous versions

That is odd, it is the same app with a different bundle id. However things could have gotten messed up since the old version uses a similar app.group container.

Please try the following:

- Make sure you deleted the other version of the app
- Ensure the proper directory is showing in the iOS app
- Re-select the directory in the iOS (choose it again after hitting the blue button)
- Make sure the new extension has the proper permissions (In settings go to Safar i > Extensions > Userscripts)
  - it should be allowed on all website always
  
If you are certain these scripts were all working in the previous version and you have checked the above, you can delete the new version and continue to use the old version that was working for you while I wait to see if this is an issue with everyone or just you.

I will **not** invalidate the old version until I am sure others are able to use the new version in the same way they used the old version.

**Comment by vjjj4688** [comment]
> @vjjj4688
> 
> Thanks for the clarification.
> 
> > It doesn't work like previous versions
> 
> That is odd, it is the same app with a different bundle id. However things could have gotten messed up since the old version uses a similar app.group container.
> 
> Please try the following:
> 
> * Make sure you deleted the other version of the app
> * Ensure the proper directory is showing in the iOS app
> * Re-select the directory in the iOS (choose it again after hitting the blue button)
> * Make sure the new extension has the proper permissions (In settings go to Safar i > Extensions > Userscripts)
>   
>   * it should be allowed on all website always
> 
> If you are certain these scripts were all working in the previous version and you have checked the above, you can delete the new version and re-install and continue to use the old version that was working for you while I wait to see if this is an issue with everyone or just you.
> 
> I will **not** invalidate the old version until I am sure others are able to use the new version in the same way they used the old version.

sorry，I used translation software, so you may have misunderstood what I meant，I mean, the last version and this version, they both do not work properly，I can see the script in the address bar of safari，But it doesn't actually run successfully，I'm sure I've given all permissions and the script is fully available on the pc
![Uploading FC2F31E2-610C-408E-83E7-35AEFAFC96DE.png…]()

**Comment by vjjj4688** [comment]
![Uploading 245238FF-78A5-4FCE-99D1-46CB07F85B38.png…]()

**Comment by quoid** [comment]
@vjjj4688 

Thank you for the additional clarification. I see you are trying to upload an image, but it does not appear to be displaying.

Since you are having an issue with both version, I suspect there could be an issue with the file you are trying to run.

As a test, please create a new userscript file with the code below and visit any website. You should see an alert show up:

```
// ==UserScript==
// @name        iosTest
// @description only for testing on ios
// @match     <all_urls>
// @run-at document-start
// @noframes
// ==/UserScript==

alert("iosTest");
```

If you do not see this alert on the first try, open the popup and ensure you see it in the list and refresh the page a couple of time.

This is will tell me if your issue is with the app or the javascript file you are trying to execute.

**Comment by vjjj4688** [comment]
> @vjjj4688
> 
> Thank you for the additional clarification. I see you are trying to upload an image, but it does not appear to be displaying.
> 
> Since you are having an issue with both version, I suspect there could be an issue with the file you are trying to run.
> 
> As a test, please create a new userscript file with the code below and visit any website. You should see an alert show up:
> 
> ```
> // ==UserScript==
> // @name        iosTest
> // @description only for testing on ios
> // @match     <all_urls>
> // @run-at document-start
> // @noframes
> // ==/UserScript==
> 
> alert("iosTest");
> ```
> 
> If you do not see this alert on the first try, open the popup and ensure you see it in the list and refresh the page a couple of time.
> 
> This is will tell me if your issue is with the app or the javascript file you are trying to execute.

After testing, it can work normally，But why can't any of the scripts I'm using myself run successfully? But why can't any of the scripts I use myself run successfully? I can only see them, but they don't work. This is the download address of the script I used，https://greasyfork.org/zh-CN/scripts/14178-ac-baidu-重定向优化百度搜狗谷歌必应搜索-favicon-双列，Maybe you can help me see if there is indeed a problem with the script? After all, I can use it properly on my PC and it's not the only one that has problems, all scripts don't work properly

**Comment by vjjj4688** [comment]
> @vjjj4688
> 
> Thank you for the additional clarification. I see you are trying to upload an image, but it does not appear to be displaying.
> 
> Since you are having an issue with both version, I suspect there could be an issue with the file you are trying to run.
> 
> As a test, please create a new userscript file with the code below and visit any website. You should see an alert show up:
> 
> ```
> // ==UserScript==
> // @name        iosTest
> // @description only for testing on ios
> // @match     <all_urls>
> // @run-at document-start
> // @noframes
> // ==/UserScript==
> 
> alert("iosTest");
> ```
> 
> If you do not see this alert on the first try, open the popup and ensure you see it in the list and refresh the page a couple of time.
> 
> This is will tell me if your issue is with the app or the javascript file you are trying to execute.

https://greasyfork.org/zh-CN/scripts/14178-ac-baidu-重定向优化百度搜狗谷歌必应搜索-favicon-双列


https://greasyfork.org/scripts/14178-ac-baidu-重定向优化百度搜狗谷歌必应搜索-favicon-双列/code/AC-baidu-重定向优化百度搜狗谷歌必应搜索_favicon_双列.user.js

**Comment by quoid** [comment]
@vjjj4688 

> After testing, it can work normally

This tells me that this is an issue with the file not the extension.

> After all, I can use it properly on my PC and it's not the only one that has problems, all scripts don't work properly

Those scripts are likely using features that are not supported with this extension or not targeting mobile versions of the website. There is a language barrier here, but the [readme documents what features are currently enabled for this extension](https://github.com/quoid/userscripts#userscripts-safari).

It's important to note that not all features present in other userscript manager are present in this userscript manager. Sometimes editing of a userscript is needed if you want to use it with this manager.

In the script you linked there is a github repository linked: https://github.com/langren1353/GM_script - it might be valuable to go there and seek help with editing the script.

**Comment by vjjj4688** [comment]
Thank you for your patience in answering，I think I know what to do.

**Comment by quoid** [comment]
If you're testing iOS and want to test macOS, please sign up here: https://github.com/quoid/userscripts/issues/164 - that platform needs more testers

I did not want to spam iOS testers with macOS invites since that's not what they signed up for, so I created a separate form.

Thanks to everyone who is helping to test!

**Comment by xinkev** [comment]
Can I take part in testing on iPadOS?

**Comment by quoid** [comment]
@xinkev sure, if you haven't already, please sign up here: https://forms.gle/QB46uYQHVyCxULue9

**Comment by ViRo3** [comment]
This [extension](https://github.com/TagoDR/MangaOnlineViewer) is very broken on iPadOS but works well on mac.

**Comment by quoid** [comment]
@ViRo3 Can you elaborate? Saying it is "very broken" doesn't offer much insight into what is going on.

Please provide your device specs, the userscript you are trying, what is the expected behaviour and what you are actually seeing. Also include steps that you took to remedy the issue, if any. Have you ensured the issue isn't caused by the userscript you are trying to execute or potentially user expectations exceeding what is possible? is it a UI issue? etc...

Myself along with others are using it on an iPad successfully, so it is worry to hear that is broken for you.

I understand you might be frustrated because something isn't working as you intend, but I need constructive feedback in order to improve the product.

**Can anyone else using this on iPad verify the "very broken" claim?**

> Would also be very helpful if it automatically opened links from userscripts sites to load into the app instead of relying on downloading files alone.

This is on the radar.

**Comment by ViRo3** [comment]
iPad 9. Extension is linked above (MangaLoader).
I absolutely didn’t mean it in any negative way and the app truly works wonders for all my extensions but the attached one above. The issue is it doesn’t load the images and UI issues are there alongwith non-interactive buttons/no animation.

it is only applicable for this extension and not any other which the app covers well. Am unaware of whether its the issue of the extension or the app.

PS : it would be very appreciated if the app automatically installed scripts in their respective pages on Greasyfork:Github too.

**Comment by quoid** [comment]
@ViRo3 

Thank you for the clarification. I hope my reply didn't seem mean, I think there was some translation issues. What you call "extension," I call a "script" or "userscript". This app is technically an app *and* extension bundled together. So when you said "This extension is very broken in iPadOS" - I understood that as "The app is very broken."

I now understand you mean that the script file you were trying was broken (or not working).

I believe the issue you are experiencing with [this script](https://github.com/TagoDR/MangaOnlineViewer/blob/master/Manga_OnlineViewer.user.js) is because several issues, and one of them is a bug (**thank you for reporting this**):

1) The script uses many external libraries that are called `@require` - however this app is **not** calling them in the proper order (that is why you can can see errors in the console, image below) - I created a new ticket for this bug here: https://github.com/quoid/userscripts/issues/166

2) This script is using some `GM` keys that are not currently supported. I do not know if this would cause the functionality to break, but it is good to know. Support for those keys will come at some point, I hope.

**Unsupported Keys**
```js
// @grant GM_getValue
// @grant GM_setValue
// @grant GM_listValues
// @grant GM_deleteValue
// @grant GM_xmlhttpRequest
```

**Error in console**

<img width="1475" alt="Screen Shot 2021-11-19 at 10 32 47 PM" src="https://user-images.githubusercontent.com/7660254/142713147-bf7c9758-238d-4364-a603-2b983219398f.png">

Thank you again for reporting this issue

**Comment by luxysiv** [comment]
https://apps.apple.com/vn/app/macaque/id1595306197?l=vi
it’s working with greasyfolk js

**Comment by Taxyovio** [comment]
Does the `@require` key work properly? 

I’ve not been able to make this [Manga Loader NSFW](https://sleazyfork.org/en/scripts/12657-manga-loader-nsfw) script work. This script is a NSFW add-on to the [main script](https://greasyfork.org/en/scripts/692-manga-loader). It uses `@require` to load the main script.

The interesting part is that the script runs properly when directly used in a bookmarketlet:
`javascript:(function(){BM_MODE=true;var addScript=function(src,onload){var s=document.createElement('script');s.onload=onload;s.src=src;document.body.appendChild(s);};addScript('https://greasyfork.org/scripts/692-manga-loader/code/Manga%20Loader.user.js',addScript.bind(null,'https://greasyfork.org/scripts/12657-manga-loader-nsfw/code/Manga%20Loader%20NSFW.user.js'));})(); `

It also works when it’s merged directly into the main script as this [gist](https://gist.github.com/Taxyovio/9848b8e533d56cbbf6405252389c92fe). 

A safe for work test manga would be https://e-hentai.org/s/5225e47182/2082483-1.

**Comment by quoid** [comment]
@Taxyovio 

Can you make a unique issue for this?

That `@require` url doesn't end with `.js` so it gets ignored. The `@require` validation doesn't account for query params. I can look into this. Can you note this in the issue you create for this?

remove the  `?29` and it would probably work

Thanks for pointing this out.

**Comment by quoid** [comment]
iOS version 1.0 has been approved and should show up on the app store soon enough. To those of you who helped test this, **I really appreciate it** and I look forward to continuing to improve the iOS experience.

Please do leave a review on the app store, especially those of you who've been using the iOS app for several weeks :) 

Thanks all, I am closing this issue now, merged https://github.com/quoid/userscripts/commit/b8b8a7ff04e8be1d56be0393a5a8369f0a566c8e
